vortex-layout/src/layouts/stats/builder.rs

```diff
@@ -0,0 +1,294 @@
+use std::marker::PhantomData;
+
+use vortex_array::arrays::ConstantArray;
+use vortex_array::builders::{ArrayBuilder, ArrayBuilderExt, BoolBuilder, builder_with_capacity};
+use vortex_array::stats::Stat;
+use vortex_array::{Array, ArrayRef, IntoArray};
+use vortex_dtype::{DType, FieldName, Nullability};
+use vortex_error::VortexResult;
+use vortex_scalar::{BinaryScalar, ScalarValue, Utf8Scalar};
+
+pub const MAX_IS_TRUNCATED: &str = "max_is_truncated";
+pub const MIN_IS_TRUNCATED: &str = "min_is_truncated";
+
+pub fn stats_builder_with_capacity(
+    stat: Stat,
+    dtype: &DType,
+    capacity: usize,
+    max_length: usize,
+) -> Box<dyn StatsArrayBuilder> {
+    let values_builder = builder_with_capacity(dtype, capacity);
+    match dtype {
+        DType::Utf8(_) | DType::Binary(_) => match stat {
+            Stat::Max => Box::new(TruncatedMaxBinaryStatsBuilder::<Utf8Scalar>::new(
+                values_builder,
+                BoolBuilder::with_capacity(Nullability::NonNullable, capacity),
+                max_length,
+            )),
+            Stat::Min => Box::new(TruncatedMinBinaryStatsBuilder::<Utf8Scalar>::new(
+                values_builder,
+                BoolBuilder::with_capacity(Nullability::NonNullable, capacity),
+                max_length,
+            )),
+            _ => Box::new(NamedArrayBuilder::new(stat, values_builder)),
+        },
+        _ => Box::new(NamedArrayBuilder::new(stat, values_builder)),
+    }
+}
+
+/// Arrays with their associated names, reduced version of a StructArray
+pub struct NamedArrays {
+    pub names: Vec<FieldName>,
+    pub arrays: Vec<ArrayRef>,
+}
+
+impl NamedArrays {
+    pub fn all_invalid(&self) -> VortexResult<bool> {
+        // by convention we assume that the first array is the one we care about for logical validity
+        self.arrays[0].all_invalid()
+    }
+}
+
+/// Minimal array builder interface for use by StatsTable for building stats arrays
+pub trait StatsArrayBuilder: Send {
```

Comment:
Why is this a trait? Vs just renaming NamedArrayBuilder -> StatsArrayBuilder?

Reply:
Because there’s stats builders that truncate. Thought that’s nicer than a switch